pkg/elasticetcd/leader.go

```diff
@@ -0,0 +1,269 @@
+package elasticetcd
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/pkg/types"
+)
+
+const (
+	eePrefix        = "elastic"
+	electionKey     = eePrefix + "/election"
+	volunteerPrefix = eePrefix + "/volunteers/"
+	nomineePrefix   = eePrefix + "/nominees/"
+	idealSizeKey    = eePrefix + "/idealSize"
+)
+
+func (ee *ElasticEtcd) startCampaign() {
+	go func() {
+		election := concurrency.NewElection(ee.session, electionKey)
+		for {
+			// Campaign for the becoming the leader
+			// Stop campaigning if the context is canceled
+			// else ignore errors and retry campaign
+			ee.log.Debug("campaigning to become leader")
+			err := election.Campaign(ee.cli.Ctx(), ee.conf.Name)
+			if err != nil {
+				switch {
+				case err == context.Canceled:
+					return
+				default:
+					continue
+				}
+			}
+			ee.log.Debug("won election to become leader")
+			// Resign as leader if you cannot start the leader funtions
+			if err := ee.startLeader(); err != nil {
+				ee.log.Debug("failed to start leader functions, resigning")
+				election.Resign(ee.cli.Ctx())
+				continue
+			}
+			// You are the leader till. So stop campaigning
+			// XXX: Do we need to check for leadership periodically, or are we the leader till we die?
+			// XXX: What happens during cluster splits? Etcd should handle that, but need to verify.
+			return
+		}
+	}()
+}
+
+func (ee *ElasticEtcd) startLeader() error {
+	ee.watchVolunteers()
+	ee.watchIdealSize()
+
+	return nil
+}
+
+func (ee *ElasticEtcd) watchVolunteers() {
+	ee.log.Debug("watching for changes to volunteers list")
+
+	f := func(_ clientv3.WatchResponse) {
+		ee.log.Debug("volunteer list had a change, doing nominations again")
+		ee.doNominations()
+	}
+
+	ee.watch(volunteerPrefix, f, clientv3.WithPrefix())
+}
+
+func (ee *ElasticEtcd) watchIdealSize() {
+	ee.log.Debug("watching for changes to ideal cluster size")
+
+	f := func(resp clientv3.WatchResponse) {
+		for _, ev := range resp.Events {
+			i, err := strconv.Atoi(string(ev.Kv.Value))
+			if err != nil {
+				ee.log.WithError(err).Error("could not parse idealsize value, ignoring update")
+				continue
+			}
+			if i != ee.conf.IdealSize {
+				ee.log.WithField("idealsize", i).Debug("idealsize changed, doing nominations again")
```

Comment:
I had forgotten about this. Right now, there isn't an API to change idealsize, but if required it can be changed directly in etcd. Everyone who can access this etcd cluster can change it right now.

We would need to protect access to the elastic namespace. Etcd v3 has users and roles, which can be used for this. I'll add a todo for this.